Add UpdateRunProgress filter for yum

Atchub opened the pull request:
@Microsoft/omsagent-devs 

Changes:
1)adds a new source for tailing yum log
2) filters the log to upload UpdateRunProgress data
3) adds a test
Test: verified with Ubuntu 16 and Redhat 6.7

Atchub commented:
@Microsoft/omsagent-devs

vrdmr commented on source/code/plugins/patch_management_lib.rb:
```diff
@@ -405,4 +406,42 @@ def process_update_run(record, tag, host, time)
+    def process_yum_update_run(record, tag, host, time)
+		list_of_updates = process_yum_record(record, tag, host, time)
+		return get_update_run_progress_blob(list_of_updates)
+		end
```
nit: Have the correct indentation.

vrdmr commented on source/code/plugins/patch_management_lib.rb:
```diff
@@ -405,4 +406,42 @@ def process_update_run(record, tag, host, time)
+			"DataItems" => list_of_updates
+		}
+		return wrapper
+	end
```
nit: Have the correct indentation.

Atchub replied:
its fixed

vrdmr commented on installer/conf/omsagent.d/patch_management.conf:
```diff
@@ -39,13 +39,31 @@
+    log_level info
+</source>
+
+<filter oms.yum_update_progress>
```
Any plans to make the changes in both package manager logs to report only the updates done by our system?

Atchub replied:
mostly next sprint but we may want to change the implementation too to depend on the runbook to get user information, yum log has that missing too

robbiezhang commented on installer/conf/omsagent.d/patch_management.conf:
```diff
@@ -27,7 +27,7 @@
 <source>
     @type tail
-    tag oms.update_progress
+    tag oms.apt_update_progress
```
i'd prefer you follow the naming convention for the tag: oms.[feature].[sub feature].

robbiezhang commented on installer/conf/omsagent.d/patch_management.conf:
```diff
@@ -39,13 +39,31 @@
+    format /(?<update-action-date>.*\d{2}:\d{2}:\d{2}) (?<update-action>.*):(?<package-name>.*)/
+    run_interval 60s
+    pos_file /var/opt/microsoft/omsagent/state/yum.log.pos
+    keep_time_key true
```
Does keep_time_key work? Since you didn't specify the time_key, it uses the default one "time". While your regex doesn't have such extraction.

robbiezhang commented on source/code/plugins/filter_linux_update_run_progress.rb:
```diff
@@ -34,7 +34,11 @@ def shutdown
     def filter(tag, time, record)
       linuxUpdates = LinuxUpdates.new(@log, @current_update_run_file)
-      return linuxUpdates.process_update_run(record, tag, @hostname, time)
+      if tag == "oms.yum_update_progress"
```
if you following the tag convention, you just need to check the last section of the tag to determine the processor.

robbiezhang commented on installer/conf/omsagent.d/patch_management.conf:
```diff
@@ -39,13 +39,31 @@
+    tag oms.yum_update_progress
+    path /var/log/yum.log
+    format /(?<update-action-date>.*\d{2}:\d{2}:\d{2}) (?<update-action>.*):(?<package-name>.*)/
+    run_interval 60s
```
run every minute?

Atchub replied:
we are trying to update the run progress ,so the lesser the latency added , the better progress update.
do you think fluentd would choke on it?

vrdmr replied:
This plugin would mostly not do any actual work most of the time. As it is tailing the yum.log, which would not be updated very frequently. I am okay with a slower time period of 5m+ as well, as it will still be reasonable time period.

robbiezhang commented on test/code/plugins/patch_management_lib_test.rb:
```diff
@@ -461,4 +461,23 @@ def test_filter_updaterun_progress
     assert_equal(fakeStartDate, result['DataItems'][0]["StartTime"])
   end
   
+    def test_filter_yum_updaterun_progress
```
nit: indent

robbiezhang commented on source/code/plugins/patch_management_lib.rb:
```diff
@@ -405,4 +406,42 @@ def process_update_run(record, tag, host, time)
+		update_run_record["TimeStamp"] = OMS::Common.format_time(time)
+		update_run_record["Tag"] = tag
+
+		update_run_record ["Start-Date"] = DateTime.parse(record["update-action-date"]).strftime(APT_GET_START_DATE_TIME_FORMAT)
```
is the update-action-date UTC time? otherwise, you need to convert it to UTC in server side.

robbiezhang commented on test/code/plugins/patch_management_lib_test.rb:
```diff
@@ -461,4 +461,23 @@ def test_filter_updaterun_progress
+    assert_equal("HostName", result['DataItems'][0]["Computer"])
+    assert_equal(@fakeUpdateRunName, result['DataItems'][0]["UpdateRunName"])
+    assert_equal('libXft-2.3.2-1.el6.x86$-1.28-12.el6.x86_64', result['DataItems'][0]["UpdateTitle"])
+    assert_equal('Succeeded', result['DataItems'][0]["Status"])
```
please add more test cases and checks:
  update-action: Installed, Erased or otherwise
  Start-Date check

Atchub commented:
fixed the comments

1) added test cases
2) removed the date as its being discussed in the team
3) fixed the naming comments and indentation
4)refactored the status and changed it to regex
